Try reconfigure

CoMPaTech opened the pull request.

bouwew commented on custom_components/plugwise/config_flow.py:
```diff
@@ -243,6 +244,66 @@ async def async_step_user(
+                await self.async_set_unique_id(
+                    api.smile_hostname or api.gateway_id, raise_on_progress=False
+                )
+                self._abort_if_unique_id_mismatch(reason="not_the_same_smile")
```
What does this do? Entering a wrong Smile ID will lead to an `invalid_auth` error.

Now that I think about it some more, only when the IP-address(, username) and Smile ID of another Plugwise Gateway are entered, then that should be signaled. This line does that?

bouwew replied:
I initially was confused by the use of MOCK_SMILE_ID in the test below.

See my feedback about this in the corresponding Core PR :)

bouwew commented:
@CoMPaTech your last commit reverses the extra logic I added to async_step_user`()`.
I changed the logic so that:
- Empty input fields are show when the user manually adds the integration. The idea was also that the input fields would show the data from discovery but I think that's not working. Maybe that can work if the logic is changed a little.
- When the user has entered data with a typo, the validation fails and the "enter data" window is shown with the data the user had entered before, including the typo, lines 233-238 should achieve that.


You should be able to keep the existing logic and add the reconfigure stuff?

CoMPaTech commented:
> @CoMPaTech your last commit reverses the extra logic I added to async_step_user`()`. I changed the logic so that:
> 
> * Empty input fields are show when the user manually adds the integration. The idea was also that the input fields would show the data from discovery but I think that's not working. Maybe that can work if the logic is changed a little.
> * When the user has entered data with a typo, the validation fails and the "enter data" window is shown with the data the user had entered before, including the typo, lines 233-238 should achieve that.
> 
> You should be able to keep the existing logic and add the reconfigure stuff?

Given the hint of suggested_values - I guess we could just solve that by rewriting the last part of the user_step to something like - saving both the 'if errors' (as it falls through anyway) and a second call to show form. We should also add a test for that to ensure we don't forget it next time tinkering. Not sure if suggested would also work for user_step though.

```
         return self.async_show_form(
             step_id=SOURCE_USER,
-            data_schema=SMILE_USER_SCHEMA(self.discovery_info),
+            data_schema=self.add_suggested_values_to_schema(
+                data_schema=SMILE_USER_SCHEMA(self.discovery_info),
+                suggested_values=self.discovery_info | user_input,
+            ),
             errors=errors,
         )
```

Thinking about the above and looking back at 233-238 - that might as well just be achievable even cleaner in the last call using   

```
-            data_schema=SMILE_USER_SCHEMA(self.discovery_info),
+           data_schema=SMILE_USER_SCHEMA(self.discovery_info | user_input)
```


But both would need some tinkering to draft out the None type handling. Or maybe better, just hand-off both discovery and user_input to SMILE_USER_SCHEMA and let it figure it out (which might be more human readable than `a | b` or `a | (b or {})` constructs (see vallox or brother takes on suggested values).

Couldn't get it to work right away, but wasn't too keen on merging this PR soon without proper cleanup anyway.

bouwew commented:
Interesting! Let's tinker a bit more during the weekend!

CoMPaTech commented:
Todo: revert + add exception as per Core 132878 - follow-up in https://github.com/plugwise/python-plugwise/issues/677

bouwew reviewed (approved):
LGTM!

bouwew commented on tests/components/plugwise/test_select.py:
```diff
@@ -86,3 +86,20 @@ async def test_legacy_anna_select_entities(
+#                ATTR_OPTION: "freezing",
+#            },
+#            blocking=True,
+#        )
```
The commented code was uncommented.
